### PR TITLE
Feature: Add type to market, order, trade and trading fees

### PR DESCRIPTION
**Pull request by dpatsora**

### CODE REVIEW

File: app/api/v2/admin/markets.rb [hunk 1]
```diff
@@ -69,7 +73,9 @@ class Markets < Grape::API
         get '/markets' do
           admin_authorize! :read, ::Market
 
-          result = ::Market.order(params[:order_by] => params[:ordering])
+          result = ::Market.where(market_type: params[:market_type])
```
Comment (dpatsora) [inline app/api/v2/admin/markets.rb]: Replace with .spot

File: app/api/v2/admin/markets.rb [hunk 2]
```diff
@@ -84,7 +90,7 @@ class Markets < Grape::API
         get '/markets/:id', requirements: { id: /[\w\.\-]+/ } do
           admin_authorize! :read, ::Market
 
-          present ::Market.find(params[:id]), with: API::V2::Admin::Entities::Market
+          present ::Market.find_spot(params[:id]), with: API::V2::Admin::Entities::Market
```
Comment (dpatsora) [inline app/api/v2/admin/markets.rb:93]: Rename market_id in controllers to market_name

File: app/api/v2/market/named_params.rb [hunk 3]
```diff
@@ -10,7 +10,7 @@ module NamedParams
         params :enabled_markets do
           requires :market,
                    type: String,
-                   values: { value: -> { ::Market.active.ids }, message: 'market.market.doesnt_exist_or_not_enabled' },
+                   values: { value: -> { ::Market.spot.active.ids }, message: 'market.market.doesnt_exist_or_not_enabled' },
```
Comment (dpatsora) [inline app/api/v2/market/named_params.rb:13]: Market names

File: app/models/market.rb [hunk 4]
```diff
@@ -122,14 +128,16 @@ class Market < ApplicationRecord
 
   # == Scopes ===============================================================
 
+  scope :spot, -> { where(market_type: 'spot') }
+  scope :otc, -> { where(market_type: 'otc') }
```
Comment (dpatsora) [inline app/models/market.rb]: Rename to qe

File: app/models/order.rb [hunk 5]
```diff
@@ -71,7 +71,7 @@ class Order < ApplicationRecord
 
   scope :done, -> { with_state(:done) }
   scope :active, -> { with_state(:wait) }
-  scope :with_market, ->(market) { where(market_id: market) }
+  scope :with_market, ->(market_name, market_type) { where(market_name: market_name, market_type: market_type) }
```
Comment (dpatsora) [inline app/models/order.rb]: Change back

File: app/api/v2/market/orders.rb [hunk 6]
```diff
@@ -137,7 +137,7 @@ class Orders < Grape::API
         params do
           optional :market,
                    type: String,
-                   values: { value: -> { ::Market.active.ids }, message: 'market.market.doesnt_exist' },
+                   values: {  value: -> { ::Market.spot.pluck(:ticker) }, message: 'market.market.doesnt_exist' },
```
Comment (dpatsora) [inline app/api/v2/market/orders.rb]: add actve

File: db/migrate/20210225123519_add_market_type.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+class AddMarketType < ActiveRecord::Migration[5.2]
+  class LegacyMarket < ActiveRecord::Base
+  end
+
+  def change
+    rename_table :markets, :legacy_markets
+
+    create_table :markets do |t|
+      t.string "ticker", limit: 20, null: false
```
Comment (calj) [inline db/migrate/20210225123519_add_market_type.rb]: let's rename it to symbol

File: app/api/v2/entities/market.rb [hunk 8]
```diff
@@ -5,11 +5,13 @@ module API
   module V2
     module Entities
       class Market < Base
+        # DEPRECATED: id became tickers
```
Comment (dpatsora) [inline app/api/v2/entities/market.rb]: Fix

File: app/api/v2/market/orders.rb [hunk 9]
```diff
@@ -137,7 +137,7 @@ class Orders < Grape::API
         params do
           optional :market,
                    type: String,
-                   values: { value: -> { ::Market.active.ids }, message: 'market.market.doesnt_exist' },
+                   values: {  value: -> { ::Market.spot.active.pluck(:symbol) }, message: 'market.market.doesnt_exist' },
                    desc: -> { V2::Entities::Market.documentation[:id] }
```
Comment (dpatsora) [inline app/api/v2/market/orders.rb]: change docs

File: app/models/trade.rb [hunk 10]
```diff
@@ -367,8 +367,9 @@ def revert_fees!
 # Indexes
 #
 #  index_trades_on_created_at                (created_at)
```
Comment (dpatsora) [inline app/models/trade.rb]: Run autoannotate

File: lib/peatio/import.rb [hunk 11]
```diff
@@ -114,7 +114,7 @@ def load_markets
       Kernel.puts 'Importing markets'
       ::Market.transaction do
         @data['markets'].map(&:symbolize_keys).each do |hash|
-          next if ::Market.exists?(id: hash.fetch(:id))
+          next if ::Market.exists?(ticker: hash.fetch(:ticker))
```
Comment (dpatsora) [inline lib/peatio/import.rb]: Change

File: spec/api/v2/admin/engines_spec.rb [hunk 12]
```diff
@@ -40,7 +40,7 @@
       expect(response).to be_successful
 
       result = JSON.parse(response.body)
-      expect(result.size).to eq 2
+      expect(result.size).to eq 3
```
Comment (dpatsora) [inline spec/api/v2/admin/engines_spec.rb:43]: The number here changed because I've added seed of QE market

File: spec/models/market_spec.rb [hunk 13]
```diff
@@ -139,8 +139,8 @@
     end
 
     it 'allows to disable all markets' do
-      Market.where.not(id: :btcusd).update_all(state: :disabled)
-      market = Market.find(:btcusd)
+      Market.where.not(ticker: :btcusd).update_all(state: :disabled)
```
Comment (dpatsora) [inline spec/models/market_spec.rb]: Update

File: spec/models/market_spec.rb [hunk 14]
```diff
@@ -207,62 +207,62 @@ def to_readable(field)
     context 'after_create' do
 
       it 'move to the bottom if there is no position' do
-        expect(Market.all.ordered.pluck(:id, :position)).to eq [["btcusd", 1], ["btceth", 2]]
+        expect(Market.spot.ordered.pluck(:ticker, :position)).to eq [["btcusd", 1], ["btceth", 2]]
```
Comment (dpatsora) [inline spec/models/market_spec.rb]: Same

File: spec/models/market_spec.rb [hunk 15]
```diff
@@ -207,62 +207,62 @@ def to_readable(field)
...
         Market.create(valid_attributes)
-        expect(Market.all.ordered.pluck(:id, :position)).to eq [["btcusd", 1], ["btceth", 2], ["btctrst", 3]]
+        expect(Market.spot.ordered.pluck(:ticker, :position)).to eq [["btcusd", 1], ["btceth", 2], ["btctrst", 3]]
```
Comment (dpatsora) [inline spec/models/market_spec.rb]: same

**Comment by dpatsora** [comment]
For fixing Market position we have separate ticket - https://openware-inc.atlassian.net/browse/NAIR-41

File: app/api/v2/admin/markets.rb [hunk 16]
```diff
@@ -84,7 +90,7 @@ class Markets < Grape::API
         get '/markets/:id', requirements: { id: /[\w\.\-]+/ } do
           admin_authorize! :read, ::Market
 
-          present ::Market.find(params[:id]), with: API::V2::Admin::Entities::Market
+          present ::Market.find_spot_by_symbol(params[:id]), with: API::V2::Admin::Entities::Market
```
Comment (mnaichuk) [inline app/api/v2/admin/markets.rb]: can be just find_spot

File: app/api/v2/entities/market.rb [hunk 17]
```diff
@@ -5,11 +5,13 @@ module API
...
       class Market < Base
+        # DEPRECATED: id became symbol
         expose(
-          :id,
+          :symbol,
+          as: :id,
           documentation: {
             type: String,
-            desc: "Unique market id. It's always in the form of xxxyyy,"\
+            desc: "Unique market ticker symbol. It's always in the form of xxxyyy,"\
```
Comment (mnaichuk) [inline app/api/v2/entities/market.rb:21]: also, add deprecated in `desc` and expose symbol without `as`

File: app/api/v2/entities/market.rb [hunk 18]
```diff
@@ -5,11 +5,13 @@ module API
...
                   "where xxx is the base currency code, yyy is the quote"\
                   "currency code, e.g. 'btcusd'. All available markets can"\
                   "be found at /api/v2/markets."
```
Comment (mnaichuk) [inline app/api/v2/entities/market.rb:24]: do we need to have in public markets entity market type?

File: app/api/v2/entities/market.rb (hunk 17 above)
Comment (dpatsora) [reply app/api/v2/entities/market.rb:21]: I think that exposing symbol instead of id will break FE

File: app/api/v2/entities/market.rb (hunk 18 above)
Comment (dpatsora) [reply app/api/v2/entities/market.rb:24]: In public no
In admin - yes

File: app/api/v2/admin/markets.rb (hunk 16 above)
Comment (dpatsora) [reply app/api/v2/admin/markets.rb]: For me *_by_symbol* is more clear

File: app/api/v2/entities/market.rb (hunk 17 above)
Comment (mnaichuk) [reply app/api/v2/entities/market.rb:21]: we should expose `both`

File: app/api/v2/admin/markets.rb [hunk 19]
```diff
@@ -53,11 +53,15 @@ class Markets < Grape::API
...
-        desc 'Get all markets, result is paginated.',
+        desc 'Get all spot or qe markets, result is paginated.',
           is_array: true,
           success: API::V2::Admin::Entities::Market
         params do
           use :pagination
+          optional :type,
+                   type: { value: String, message: 'admin.market.non_string_market_type' },
+                   values: { value: -> { ::Market::TYPES }, message: 'admin.market.invalid_market_type' },
+                   default: 'spot'
```
Comment (calj) [inline app/api/v2/admin/markets.rb]: You can define a ::Market::DEFAULT_TYPE

File: app/api/v2/admin/markets.rb [hunk 20]
```diff
@@ -53,11 +53,15 @@ class Markets < Grape::API
           end
         end
 
-        desc 'Get all markets, result is paginated.',
+        desc 'Get all spot or qe markets, result is paginated.',
```
Comment (calj) [inline app/api/v2/admin/markets.rb]: Let's not list types in description

File: app/api/v2/admin/markets.rb [hunk 21]
```diff
@@ -69,22 +73,28 @@ class Markets < Grape::API
...
         params do
-          requires :id,
+          requires :symbol,
                    type: String,
-                   desc: -> { API::V2::Admin::Entities::Market.documentation[:id][:desc] }
+                   desc: -> { API::V2::Admin::Entities::Market.documentation[:symbol][:desc] }
+          optional :type,
+                   type: { value: String, message: 'admin.market.non_string_market_type' },
+                   values: { value: -> { ::Market::TYPES }, message: 'admin.market.invalid_market_type' },
+                   default: 'spot'
```
Comment (calj) [inline app/api/v2/admin/markets.rb]: use everywhere the default market type

File: db/migrate/20210225123519_add_market_type.rb [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+    drop_table :legacy_markets
+
+    add_column(:orders, :market_type, :string, null: false, default: 'spot', after: 'market_id')
+    remove_index(:orders, %w[type market_id]) if index_exists?(:orders, %w[type market_id])
+    remove_index(:orders, %w[type state market_id]) if index_exists?(:orders, %w[type state market_id])
+    add_index(:orders, %w[type market_id market_type]) unless index_exists?(:orders, %w[type market_id market_type])
+    add_index(:orders, %w[type state market_id market_type]) unless index_exists?(:orders, %w[type state market_id market_type])
+
+
+    add_column(:trades, :market_type, :string, null: false, default: 'spot', after: 'market_id')
```
Comment (calj) [inline db/migrate/20210225123519_add_market_type.rb:49]: let's add the market_type to fees_schedules, default it to spot.
also add the market_type to fees_schedules endpoints.

File: app/api/v2/admin/markets.rb [hunk 23]
```diff
@@ -135,8 +149,13 @@ class Markets < Grape::API
         end
         params do
           use :update_market_params
+          # Id parameter should be deprecated and changed to symbol
           requires :id,
```
Comment (mnaichuk) [inline app/api/v2/admin/markets.rb]: as: :symbol and you can remove `except` below

Comment (mnaichuk) [reply app/api/v2/admin/markets.rb]: and let's add optional `symbol` with exactly_one_of

File: app/api/v2/admin/orders.rb [hunk 24]
```diff
@@ -77,23 +81,24 @@ class Orders < Grape::API
         post '/orders/:id/cancel' do
           admin_authorize! :update, ::Order
 
+          # This part won't work with Finex
```
Comment (mnaichuk) [inline app/api/v2/admin/orders.rb:84]: lets make it work:)

File: app/api/v2/admin/orders.rb [hunk 25]
```diff
@@ -77,23 +81,24 @@ class Orders < Grape::API
...
           begin
             order = Order.find(params[:id])
             order.trigger_cancellation
             present order, with: API::V2::Admin::Entities::Order
           rescue ActiveRecord::RecordNotFound => e
             # RecordNotFound in rescued by ExceptionsHandler.
             raise(e)
-          rescue
+          rescue => e
```
Comment (mnaichuk) [inline app/api/v2/admin/orders.rb]: lets rescue StandardError

File: app/api/v2/management/markets.rb [hunk 26]
```diff
@@ -12,12 +12,17 @@ class Markets < Grape::API
         end
 
         params do
+          # Id parameter should be deprecated and changed to symbol
           requires :id,
```
Comment (mnaichuk) [inline app/api/v2/management/markets.rb:94]: same as in admin api

File: app/api/v2/market/orders.rb [hunk 27]
```diff
@@ -11,16 +11,21 @@ class Orders < Grape::API
           is_array: true,
           success: API::V2::Entities::Order
         params do
+          # Would be cool to validate market based on the market_type
```
Comment (mnaichuk) [inline app/api/v2/market/orders.rb]: better to do `TODO` format

File: db/migrate/20210225123519_add_market_type.rb [hunk 28]
```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+class AddMarketType < ActiveRecord::Migration[5.2]
+  class LegacyMarket < ActiveRecord::Base
+  end
+
+  def change
```
Comment (mnaichuk) [inline db/migrate/20210225123519_add_market_type.rb]: lets add dir up | down

File: app/models/trading_fee.rb [hunk 29]
```diff
@@ -68,12 +68,19 @@ class TradingFee < ApplicationRecord
...
+            presence: true,
+            inclusion: { in: ->(_fs) { Market::TYPES } }
 
   validates :maker, :taker, precision: { less_than_or_eq_to: FEE_PRECISION }
 
   # == Scopes ===============================================================
 
+  scope :spot, -> { where(market_type: 'spot') }
+  scope :qe, -> { where(market_type: 'qe') }
+
```
Comment (mnaichuk) [inline app/models/trading_fee.rb:83]: update `for` function with `market_type` param

File: app/api/v2/admin/markets.rb (hunk 23 above)
Comment (dpatsora) [reply app/api/v2/admin/markets.rb]: I faced : {errors: "exclusive"}

File: db/migrate/20210225123519_add_market_type.rb (hunk 28 above)
Comment (dpatsora) [reply db/migrate/20210225123519_add_market_type.rb]: Please, check now

**Pull request merged**